### PR TITLE
Add new rule 'default-props-match-prop-types'

### LINKED ISSUE #1022: New rule: `no-invalid-default-props`
I'd like a rule that errors if anything is in `defaultProps` that is not also in `propTypes`.

This came up with the following code:
```js
Component.propTypes = {
  fooValue: PropTypes.string,
};
Component.defaultProps = {
  foo: '',
};
```

In other words, it was a partially completed prop rename.

This rule should also look up a propTypes variable (`Component.propTypes = propTypes`) defined elsewhere in the file.

### PR DESCRIPTION
**Pull request by webOS101**
Fixes #1022 

There's currently one caveat that I haven't addressed: If there are **no** `propTypes` at all no warnings are issued. I didn't distinguish between that case and the case where they are not able to be determined (e.g.: external import).  Also, this rule may conflict with `require-default-props` as I am re-using the same component members, though testing did not reveal any problems using the two rules together.

### CODE REVIEW

File: lib/rules/no-invalid-default-props.js [hunk 1]
```diff
@@ -0,0 +1,624 @@
...
+     * @returns {Boolean}       `true` if the node is a propTypes declaration, `false` if not
+     */
+    function isPropTypesDeclaration(node) {
+      return getPropertyName(node) === 'propTypes';
+    }
+
+    /**
+     * Checks if the Identifier node passed in looks like a defaultProps declaration.
+     * @param   {ASTNode}  node The node to check. Must be an Identifier node.
+     * @returns {Boolean}       `true` if the node is a defaultProps declaration, `false` if not
```
Comment (ljharb) [inline lib/rules/no-invalid-default-props.js]: I don't think any of these spaces are needed; do we typically visually align these comments in other rules?

**Comment by lencioni** [comment]
Thanks for the contribution! I haven't had time to review the code here yet, but I wanted to mention that I think `no-invalid-default-props` isn't quite the right name for this rule. To me that name implies that the default props are going to be checked against the propTypes validators.

I'm not quite sure what a good name might be. Maybe something along the lines of `no-extra-default-props`?

Also, I'm not entirely sure this makes sense as a separate rule instead of enhancing the `prop-types` rule to include this. @ljharb what do you think?

File: lib/rules/no-invalid-default-props.js [hunk 2]
```diff
@@ -0,0 +1,624 @@
...
+     */
+    function isDefaultPropsDeclaration(node) {
+      return (getPropertyName(node) === 'defaultProps' || getPropertyName(node) === 'getDefaultProps');
```
Comment (ljharb) [inline lib/rules/no-invalid-default-props.js]: Let's cache the property name rather than calling the function twice.

File: lib/rules/no-invalid-default-props.js [hunk 3]
```diff
@@ -0,0 +1,624 @@
...
+          } else {
+            addDefaultPropsToComponent(component, getDefaultPropsFromObjectExpression(expression));
+          }
+
+          return;
+        }
+
+        // e.g.:
+        // MyComponent.propTypes.baz = React.PropTypes.string;
+        if (node.parent.type === 'MemberExpression' && node.parent.parent.type === 'AssignmentExpression') {
```
Comment (ljharb) [inline lib/rules/no-invalid-default-props.js]: Will `node.parent` always have a `.parent` node?

File: lib/rules/no-invalid-default-props.js [hunk 4]
```diff
@@ -0,0 +1,624 @@
...
+
+        if (!node.static) {
+          return;
+        }
+
+        if (!node.value) {
+          return;
+        }
+
+        var isPropType = getPropertyName(node) === 'propTypes';
```
Comment (ljharb) [inline lib/rules/no-invalid-default-props.js]: Let's cache the property name here rather than calling the function thrice.

**Comment by ljharb** [comment]
Hmm, that makes sense that "invalid" isn't the best name. `extra` works, I suppose - or `no-superfluous-default-props`, or some other synonym?

**Comment by ljharb** [comment]
I do think it belongs as a separate rule since it warns on defaults, not prop types.

File: lib/rules/no-invalid-default-props.js (hunk 1 above)
Comment (webOS101) [inline lib/rules/no-invalid-default-props.js]: Happy to make changes, but this code is nearly identical to `require-default-props`.  Do you want me to update the source there, too?

**Comment by webOS101** [comment]
`no-unnecessary-default-props`?  `no-unneeded-default-props`?  `no-useless-default-props`?

File: lib/rules/no-invalid-default-props.js (hunk 1 above)
Comment (ljharb) [inline lib/rules/no-invalid-default-props.js]: in that case (that our rules already are inconsistent) then don't worry about it; it can be cleaned up in a separate PR.

**Comment by ljharb** [comment]
A defaultProp for a required prop is arguably not "useless" since it may serve as documentation (not that I would ever use the option that allows it) so I don't think that's the right word.

`unnecessary` could work too (along with `extra` or `superfluous`).

`no-unused-default-props` would be good but there's no safe way to tell if the *user* uses it, only if React will use it.

Let's keep bikeshedding on the name, might as well pick the best one now.

File: lib/rules/no-invalid-default-props.js (hunk 3 above)
Comment (webOS101) [inline lib/rules/no-invalid-default-props.js]: I think it's unlikely you wouldn't have a parent node in the case of a `MemberExpression` whose parent is a `MemberExpression` and belongs to a Component.  But, I will admit I'm not intimately familiar with this code.

Comment (ljharb) [inline lib/rules/no-invalid-default-props.js]: I don't know much of anything about AST nodes, I just want to avoid a crash in the future when a new node type comes out

**Comment by webOS101** [comment]
@lencioni A rule that validates the types of the `defaultProps` seems very useful!

**Comment by webOS101** [comment]
I addressed some of the concerns raised in the review, but understand we're still searching for a good name.

**Comment by webOS101** [comment]
Anyone had any good ideas on the naming?  Nothing particularly new has come to me.  Maybe `no-stray-default-props`?

**Comment by ljharb** [comment]
What about `default-props-match-prop-types`?

**Comment by webOS101** [comment]
That seems more in keeping with @lencioni 's checker that would validate types.  `no-undefined-default-props`?  `no-unexpected-default-props`?  `only-meaningful-default-props`?

**Comment by ljharb** [comment]
I think `default-props-match-prop-types` is good, and checking for matching types would be enabled in that rule via an additional option :-)

**Comment by webOS101** [comment]
I'm curious how accurate we could be in matching the types.  Certainly simple types would be easy, more complicated ones could easily become problematic.  If you like the new name a lot, I can update the PR to the new name.

**Comment by ljharb** [comment]
I think matching the types will likely only be possible when using the built-in `PropTypes` validators, or else it'd have to actually pull in and evaluate any custom prop types - but that's a discussion for that separate PR.

I think "match" doesn't necessarily connote "types match", which is why I don't think having the option to match types is a blocker. I mainly just want to get this in, and I think that's a good enough name ¯\\\_(ツ)_/¯

**Comment by webOS101** [comment]
Updated the name of the rule and rebased on master.

File: lib/rules/default-props-match-prop-types.js [hunk 5]
```diff
@@ -0,0 +1,627 @@
+/**
+ * @fileOverview Enforce all defaultProps are defined in propTypes
+ * @author Vitor Balocco
+ * @author Roy Sutton
+ */
+'use strict';
+
+var has = require('has');
+var find = require('array.prototype.find');
```
Comment (ljharb) [inline lib/rules/default-props-match-prop-types.js]: this is no longer necessary, since master now only supports node 4+

File: lib/rules/default-props-match-prop-types.js [hunk 6]
```diff
@@ -0,0 +1,627 @@
...
+      return propTypeExpression.type === 'MemberExpression' && propTypeExpression.property.name === 'isRequired';
+    }
+
+    /**
+     * Find a variable by name in the current scope.
+     * @param  {string} name Name of the variable to look for.
+     * @returns {ASTNode|null} Return null if the variable could not be found, ASTNode otherwise.
+     */
+    function findVariableByName(name) {
+      var variable = find(variableUtil.variablesInScope(context), function(item) {
```
Comment (ljharb) [inline lib/rules/default-props-match-prop-types.js]: so this can be `const variable = variableUtil.variablesInScope(context).find(item => item.name === name);`

File: lib/rules/default-props-match-prop-types.js [hunk 7]
```diff
@@ -0,0 +1,627 @@
...
+
+    /**
+     * Extracts a DefaultProp from an ObjectExpression node.
+     * @param   {ASTNode} objectExpression ObjectExpression node.
+     * @returns {Object|string}            Object representation of a defaultProp, to be consumed by
+     *                                     `addDefaultPropsToComponent`, or string "unresolved", if the defaultProps
+     *                                     from this ObjectExpression can't be resolved.
+     */
+    function getDefaultPropsFromObjectExpression(objectExpression) {
+      var hasSpread = find(objectExpression.properties, function(property) {
```
Comment (ljharb) [inline lib/rules/default-props-match-prop-types.js]: and this

File: lib/rules/default-props-match-prop-types.js [hunk 8]
```diff
@@ -0,0 +1,627 @@
...
+
+      addPropTypesToComponent(component, getPropTypesFromTypeAnnotation(node.typeAnnotation, context));
+    }
+
+    function isPropTypeAnnotation(node) {
+      return (getPropertyName(node) === 'props' && !!node.typeAnnotation);
+    }
+
+    function propFromName(propTypes, name) {
+      return find(propTypes, function (prop) {
```
Comment (ljharb) [inline lib/rules/default-props-match-prop-types.js]: and this can be `return propTypes.find(prop => prop.name === name);`

File: lib/rules/default-props-match-prop-types.js (hunk 5 above)
Comment (webOS101) [inline lib/rules/default-props-match-prop-types.js]: What's sad is I originally implemented with `Array.find()` but realized this didn't match with the current style.

**Comment by webOS101** [comment]
Updated all the `var` declarations as well.

**Comment by webOS101** [comment]
Anything else needed for this?

**Comment by ljharb** [comment]
Sorry for the delay. I'll give this another day or so, and if there's no objections, I'll merge it.

**Comment by webOS101** [comment]
Whoops, looks like I didn't add the rule to the README.md in the root.  Should I create a separate PR for that?

**Comment by ljharb** [comment]
@webOS101 yes, thanks, that'd be great!

**Comment by webOS101** [comment]
https://github.com/yannickcr/eslint-plugin-react/pull/1239
